{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5cb28939",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import Entrez\n",
    "Entrez.email = 'A.N.Other@example.com'\n",
    "\n",
    "import Bio\n",
    "import mygene\n",
    "import csv\n",
    "import re\n",
    "import datetime\n",
    "import os\n",
    "import requests\n",
    "\n",
    "from tqdm import tqdm\n",
    "from Bio.Seq import Seq\n",
    "from Bio import SeqIO\n",
    "from Bio.SeqFeature import SeqFeature, FeatureLocation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a015b373",
   "metadata": {},
   "outputs": [],
   "source": [
    "# TargetSeqLibrary class contains methods which construct a library of valid genes and their target sequences in the form of a CSV file.\n",
    "# Genes which are not found in the NCBI database or inputted PAM is not found in select exons of a gene are added to a library of invalid genes in the form of a CSV file.\n",
    "class TargetSeqLibrary:\n",
    "\n",
    "    # Constructor for TargetSeqLibrary\n",
    "    def __init__(self, input_CSV, input_species, input_PAM_seq, input_repeat_seq, input_length_of_target_seq, output_valid_CSV_file_path, output_invalid_CSV_file_path):\n",
    "        self.CSV_of_genes = open(input_CSV, 'r')\n",
    "        self.species = input_species\n",
    "        self.PAM = input_PAM_seq\n",
    "        self.repeat_seq = input_repeat_seq\n",
    "        self.length_target_seq = input_length_of_target_seq\n",
    "        self.valid_target_seq_CSV_path = output_valid_CSV_file_path\n",
    "        self.invalid_gene_CSV_path = output_invalid_CSV_file_path\n",
    "        \n",
    "        self.valid_gene_dict = {}\n",
    "        self.invalid_gene_dict = {'gene_ids': []}\n",
    "    \n",
    "    # Query gene accession numbers and exon ranges from NCBI for each gene in inputted CSV file\n",
    "    def query_accession_numbers(self):\n",
    "        mg = mygene.MyGeneInfo()\n",
    "        exon_list = mg.querymany(self.CSV_of_genes, scopes = 'symbol', fields = 'exons', species = self.species, returnall = True)\n",
    "        self.add_not_found_in_database_gene_to_target_seqs(exon_list['missing'])\n",
    "        return exon_list\n",
    "    \n",
    "    # Add gene with valid target sequence into valid_gene_dict\n",
    "    def add_gene_to_found_seqs(self, first_variant_of_found_gene):\n",
    "        self.valid_gene_dict.update({first_variant_of_found_gene.gene_id : {'gRNA':first_variant_of_found_gene.return_gRNA_seq()}})\n",
    "        \n",
    "    # Add gene where suitable target sequence not found in select exons into invalid_gene_dict\n",
    "    def add_not_found_target_seq_gene_to_target_seqs(self, first_variant_of_not_found_gene):\n",
    "        if first_variant_of_not_found_gene.gene_id not in self.invalid_gene_dict['gene_ids']:\n",
    "            self.invalid_gene_dict['gene_ids'].append(first_variant_of_not_found_gene.gene_id)\n",
    "        \n",
    "    # Add gene which was not found in NCBI database into invalid_gene_dict\n",
    "    def add_not_found_in_database_gene_to_target_seqs(self, list_of_not_found_genes):\n",
    "        for gene_id in list_of_not_found_genes:\n",
    "            if gene_id not in self.invalid_gene_dict['gene_ids']:\n",
    "                self.invalid_gene_dict['gene_ids'].append(gene_id)\n",
    "    \n",
    "    # Iterate through exon_list attained from NCBI, check if each gene contains a suitable target sequence, and add gene sequence to respective dictionary\n",
    "    def check_every_gene_for_target_seq(self):\n",
    "        # Wrap the loop with tqdm to display the progress bar\n",
    "        for gene in tqdm(self.query_accession_numbers()['out'], total=len(self.query_accession_numbers()['out'])):\n",
    "            if 'exons' in gene:\n",
    "                all_exon_ranges = gene['exons']\n",
    "                gene_id = gene['query']\n",
    "                accession_number_of_first_variant = (all_exon_ranges[0])['transcript']\n",
    "                first_variant_strand_directionality = (all_exon_ranges[0])['strand']\n",
    "                first_variant = SingleGeneSeq(gene_id, accession_number_of_first_variant, first_variant_strand_directionality, self.PAM, self.repeat_seq, self.length_target_seq)\n",
    "\n",
    "                target_seq = first_variant.search_for_PAM()\n",
    "                \n",
    "                # Check if a valid target sequence was found and if the length of found gRNA is the same as the inputted length of the target sequence plus the length of the inputted repeat sequence.\n",
    "                # The second condition may occur if the PAM is found too close to the start or end of an exon, meaning that the target sequence will be shorter than the inputted length of the target sequence.\n",
    "                if isinstance(target_seq, Bio.SeqRecord.SeqRecord) and first_variant.return_length_of_gRNA_seq() == (self.length_target_seq + len(self.repeat_seq)) :\n",
    "                    self.add_gene_to_found_seqs(first_variant)\n",
    "                    \n",
    "                else: \n",
    "                    if first_variant.gene_id not in self.invalid_gene_dict['gene_ids']:\n",
    "                        self.add_not_found_target_seq_gene_to_target_seqs(first_variant)\n",
    "\n",
    "        self.save_CSV_of_valid_genes(self.valid_target_seq_CSV_path)\n",
    "        self.save_CSV_of_invalid_genes(self.invalid_gene_CSV_path)\n",
    "        \n",
    "    # Write valid_gene_dict into a CSV \n",
    "    def save_CSV_of_valid_genes(self, file_path):\n",
    "        try:\n",
    "            with open(file_path, 'w', newline='') as file:\n",
    "                writer = csv.writer(file)\n",
    "                writer.writerow(['gene id', 'gRNA'])\n",
    "                for gene_id, values in self.valid_gene_dict.items():\n",
    "                    writer.writerow([gene_id, values['gRNA']])\n",
    "        except IOError:\n",
    "            print(\"An error occurred while writing to a CSV file.\")\n",
    "    \n",
    "    # Write invalid_gene_dict into a CSV      \n",
    "    def save_CSV_of_invalid_genes(self, file_path):\n",
    "        try:\n",
    "            with open(file_path, 'w', newline='') as file:\n",
    "                writer = csv.writer(file)\n",
    "                writer.writerow(['gene_id'])\n",
    "                for gene_id in self.invalid_gene_dict['gene_ids']:\n",
    "                    writer.writerow([gene_id])\n",
    "        except IOError:\n",
    "            print(\"An error occurred while writing to a CSV file.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "4ab90475",
   "metadata": {},
   "outputs": [],
   "source": [
    "# SingleGeneSeq class contains methods which construct a gene and find a valid target sequence for the gene's first variant.\n",
    "class SingleGeneSeq:\n",
    "  \n",
    "    # Constructor for SingleGeneSeq\n",
    "    def __init__(self, gene_id, accession_number_of_first_variant, first_variant_strand_directionality, input_PAM_seq, input_repeat_seq, input_length_of_target_seq):\n",
    "        self.gene_id = gene_id\n",
    "        self.accession_number_of_first_variant = accession_number_of_first_variant\n",
    "        self.first_variant_strand_directionality = first_variant_strand_directionality\n",
    "        self.input_PAM_seq = input_PAM_seq\n",
    "        self.input_repeat_seq = input_repeat_seq\n",
    "        self.input_length_of_target_seq = input_length_of_target_seq\n",
    "        \n",
    "        self.record = None\n",
    "        self.start_pos = None\n",
    "        self.prev_pos_of_PAM = 0\n",
    "        self.seq_to_search = None\n",
    "        self.range_of_first_exon_with_start_of_CDS = None\n",
    "        self.target_seq = None\n",
    "        self.target_seq_length = 0\n",
    "        self.start_direction_of_PAM_search = None\n",
    "        self.start_index = 0\n",
    "        self.end_of_CDS_range_in_variant = 0\n",
    "        \n",
    "    # Retrieve details from NCBI based on a gene's first variant. These details include the exon and coding sequence range specific to the first variant.\n",
    "    def fetch_details(self, accession_number_of_first_variant):\n",
    "        handle = Entrez.efetch(db = \"nucleotide\", id = accession_number_of_first_variant, rettype = \"gb\", retmode = \"text\")\n",
    "        details = SeqIO.read(handle, \"genbank\")\n",
    "        handle.close()\n",
    "        return details\n",
    " \n",
    "    # Find first exon range which includes the beginning of the coding sequence\n",
    "    def find_range_of_seq_to_search_with_CDS(self):\n",
    "        start_of_CDS_range_of_variant = None\n",
    "        exon_range_of_variant = None\n",
    "        \n",
    "        self.record = self.fetch_details(self.accession_number_of_first_variant)\n",
    "        \n",
    "        # Iterating through features of variant to find range of coding strand\n",
    "        for feat in self.record.features:\n",
    "            if feat.type == 'CDS':\n",
    "                start_of_CDS_range_of_variant = feat.location.start + 1\n",
    "                self.end_of_CDS_range_in_variant = feat.location.end\n",
    "                break\n",
    "            \n",
    "        if start_of_CDS_range_of_variant is None or self.end_of_CDS_range_in_variant is None:\n",
    "            return None\n",
    "\n",
    "        # Iterating through features of variant to find range of exon which contains start of coding strand\n",
    "        for feat in self.record.features:\n",
    "            if feat.type == 'exon' and feat.location.end > start_of_CDS_range_of_variant:\n",
    "                exon_range_of_variant = feat.location\n",
    "                break\n",
    "            \n",
    "        if exon_range_of_variant != None:\n",
    "            # Check that end of exon range is within the range of the CDS\n",
    "            if exon_range_of_variant.end > self.end_of_CDS_range_in_variant:\n",
    "                self.range_of_first_exon_with_start_of_CDS = FeatureLocation(start_of_CDS_range_of_variant, self.end_of_CDS_range_in_variant)\n",
    "                return self.range_of_first_exon_with_start_of_CDS\n",
    "            \n",
    "            self.range_of_first_exon_with_start_of_CDS = FeatureLocation(start_of_CDS_range_of_variant, exon_range_of_variant.end)\n",
    "            return self.range_of_first_exon_with_start_of_CDS\n",
    "        \n",
    "        else:\n",
    "            return None\n",
    "    \n",
    "    \n",
    "    def find_range_of_seq_to_search_past_CDS(self):\n",
    "        next_exon_range_of_variant = None\n",
    "        self.record = self.fetch_details(self.accession_number_of_first_variant)\n",
    "        \n",
    "        for feat in self.record.features:\n",
    "            # Due to the transition from 1-based indexing of the NCBI database to 0-based indexing of Python for a \n",
    "            # FeatureLocation object, add 1 to the starting position of an exon to account for this when comparing two FeatureLocations.\n",
    "            if feat.type == 'exon' and (feat.location.start + 1) > self.range_of_first_exon_with_start_of_CDS.end:\n",
    "                next_exon_range_of_variant = feat.location\n",
    "                break\n",
    "\n",
    "        if next_exon_range_of_variant != None and next_exon_range_of_variant.start < self.end_of_CDS_range_in_variant:\n",
    "            if next_exon_range_of_variant.end > self.end_of_CDS_range_in_variant:\n",
    "                next_exon_location = FeatureLocation(next_exon_range_of_variant.start, self.end_of_CDS_range_in_variant)\n",
    "                return next_exon_location\n",
    "            \n",
    "            next_exon_location = FeatureLocation(next_exon_range_of_variant.start, next_exon_range_of_variant.end)\n",
    "            return next_exon_location\n",
    "        \n",
    "        else:\n",
    "            return None\n",
    "    \n",
    "    # Extract the sequence of a specific range to look for PAM.\n",
    "    def retrieve_seq_to_search_for_PAM(self):\n",
    "        feat_location = self.find_range_of_seq_to_search_with_CDS()\n",
    "        self.range_of_first_exon_with_start_of_CDS = feat_location\n",
    "\n",
    "        if feat_location is None:\n",
    "            return None\n",
    "        \n",
    "        # Extracting the selected exon sequence\n",
    "        exon_seq_to_extract = SeqFeature(feat_location)\n",
    "        self.seq_to_search = exon_seq_to_extract.extract(self.fetch_details(self.accession_number_of_first_variant)) \n",
    "        \n",
    "        if self.seq_to_search is None:\n",
    "            return None \n",
    "            \n",
    "        return self.seq_to_search\n",
    "\n",
    "    def search_for_PAM(self):\n",
    "        PAM = self.input_PAM_seq\n",
    "        \n",
    "        # Searching for PAM from left of exon sequence\n",
    "        CDS_exon_seq = self.retrieve_seq_to_search_for_PAM()\n",
    "        \n",
    "        if isinstance(self.seq_to_search, Bio.SeqRecord.SeqRecord):\n",
    "            self.seq_to_search = CDS_exon_seq\n",
    "            self.start_index = 0\n",
    "            \n",
    "            while True:\n",
    "                self.start_pos = str(self.seq_to_search.seq[self.start_index:]).find(PAM)\n",
    "                self.start_direction_of_PAM_search = \"left_to_right\"\n",
    "                self.target_seq = self.find_target_seq()\n",
    "                \n",
    "                if self.start_pos == -1:\n",
    "                    break\n",
    "              \n",
    "                if len(self.target_seq) == self.input_length_of_target_seq:\n",
    "                    self.start_pos = self.start_pos + self.start_index\n",
    "                    self.target_seq = self.find_target_seq()\n",
    "                    break\n",
    "                    \n",
    "                self.start_index = self.start_index + self.start_pos + len(PAM)            \n",
    "                \n",
    "            # If PAM not found, search for reverse complement of PAM from left of exon sequence\n",
    "            if self.start_pos == -1 or len(self.find_target_seq()) != self.input_length_of_target_seq:\n",
    "                self.start_index = 0\n",
    "                complement_PAM = Seq(PAM).complement()\n",
    "                reversed_complement_PAM = str(complement_PAM[::-1])\n",
    "                self.prev_pos_of_PAM = self.start_index\n",
    "\n",
    "                while True:\n",
    "                    self.start_pos = str(self.seq_to_search.seq[self.start_index:]).find(reversed_complement_PAM)\n",
    "                    self.start_direction_of_PAM_search = \"right_to_left\"\n",
    "                    self.target_seq = self.find_target_seq()\n",
    "                    \n",
    "                    if self.start_pos == -1:\n",
    "                        break\n",
    "                        \n",
    "                    if len(self.target_seq) == self.input_length_of_target_seq:\n",
    "                        self.start_pos = self.start_pos + self.start_index\n",
    "                        self.target_seq = self.find_target_seq()\n",
    "                        break\n",
    "                        \n",
    "                    else:\n",
    "                        self.start_index = self.start_index + self.start_pos + len(reversed_complement_PAM)\n",
    "\n",
    "                # Search next exon (if exists) if reverse complement of PAM not found\n",
    "                if self.start_pos == -1 or len(self.find_target_seq()) != self.input_length_of_target_seq:\n",
    "                    self.start_index = 0\n",
    "                    self.prev_pos_of_PAM = self.start_pos\n",
    "                    next_exon_location = self.find_range_of_seq_to_search_past_CDS()\n",
    "\n",
    "                    # Extracting the next exon sequence\n",
    "                    if next_exon_location != None:\n",
    "                        next_exon_seq_to_extract = SeqFeature(next_exon_location)\n",
    "                        self.seq_to_search = next_exon_seq_to_extract.extract(self.record)\n",
    "                        \n",
    "                    else:\n",
    "                        return None\n",
    "\n",
    "                    # Searching for PAM from left of next exon sequence\n",
    "                    while True:\n",
    "                        self.start_pos = str(self.seq_to_search.seq[self.start_index:]).find(PAM)\n",
    "                        self.start_direction_of_PAM_search = \"left_to_right\"\n",
    "                        self.target_seq = self.find_target_seq()\n",
    "                        \n",
    "                        if self.start_pos == -1:\n",
    "                            break\n",
    "                            \n",
    "                        if len(self.target_seq) == self.input_length_of_target_seq:\n",
    "                            self.start_pos = self.start_pos + self.start_index\n",
    "                            self.target_seq = self.find_target_seq()\n",
    "                            break\n",
    "                                \n",
    "                        self.start_index = self.start_index + self.start_pos + len(PAM)\n",
    "\n",
    "                    # Search this same exon for reverse complement of PAM from left of exon sequence\n",
    "                    if self.start_pos == -1 or len(self.find_target_seq()) != self.input_length_of_target_seq:\n",
    "                        self.start_index = 0\n",
    "                        \n",
    "                        while True:\n",
    "                            self.start_index = self.start_index + self.start_pos + len(reversed_complement_PAM)\n",
    "                            self.prev_pos_of_PAM = self.start_index\n",
    "                            self.start_pos = str(self.seq_to_search.seq[self.start_index:]).find(reversed_complement_PAM)\n",
    "                            self.start_direction_of_PAM_search = \"right_to_left\"\n",
    "    \n",
    "                            self.target_seq = self.find_target_seq()\n",
    "                                \n",
    "                            if self.start_pos == -1:\n",
    "                                break\n",
    "                            \n",
    "                            if len(self.target_seq) == self.input_length_of_target_seq:\n",
    "                                self.start_pos = self.start_pos + self.start_index\n",
    "                                break  \n",
    "                      \n",
    "                        if self.start_pos == -1 or len(self.target_seq) != self.input_length_of_target_seq:\n",
    "                            return None\n",
    "            \n",
    "            return self.seq_to_search\n",
    "        \n",
    "        return None          \n",
    "\n",
    "    # Cleave DNA from end/beginning of PAM to a given number of nucleotides past/before the PAM\n",
    "    def find_target_seq(self):\n",
    "        if self.start_pos != -1 and isinstance(self.seq_to_search, Bio.SeqRecord.SeqRecord):\n",
    "            \n",
    "            if self.start_direction_of_PAM_search == \"left_to_right\":\n",
    "                target_seq_DNA = self.seq_to_search.seq[(self.start_pos + len(self.input_PAM_seq)) : (self.start_pos + len(self.input_PAM_seq) + self.input_length_of_target_seq)]\n",
    "                return target_seq_DNA\n",
    "            \n",
    "            elif self.start_direction_of_PAM_search == \"right_to_left\":\n",
    "                start_of_target_seq = (self.prev_pos_of_PAM + self.start_pos) - self.input_length_of_target_seq\n",
    "                target_seq_DNA = (self.seq_to_search.seq[(start_of_target_seq) : (self.prev_pos_of_PAM + self.start_pos)])[::-1]\n",
    "                return target_seq_DNA \n",
    "        \n",
    "        return None\n",
    "    \n",
    "    # Transcribe the target DNA sequence to RNA\n",
    "    def transcribe_to_RNA(self):\n",
    "        if isinstance(self.target_seq, Bio.Seq.Seq):\n",
    "            target_seq_RNA = ((self.target_seq).transcribe())\n",
    "            return target_seq_RNA\n",
    "        \n",
    "        return None\n",
    "    \n",
    "    # Prepend the repeat sequence to the target RNA sequence, thereby creating gRNA \n",
    "    def add_repeat_seq(self):\n",
    "        target_RNA_seq = self.transcribe_to_RNA()\n",
    "        if target_RNA_seq != None:\n",
    "            gRNA = self.input_repeat_seq + target_RNA_seq\n",
    "            return gRNA\n",
    "        \n",
    "        return None\n",
    "    \n",
    "    # Return gRNA sequence\n",
    "    def return_gRNA_seq(self):\n",
    "        if self.add_repeat_seq() != None:\n",
    "            return self.add_repeat_seq()\n",
    "        \n",
    "        return None\n",
    "    \n",
    "    # Return length of gRNA sequence\n",
    "    def return_length_of_gRNA_seq(self):\n",
    "        if self.return_gRNA_seq() != None:\n",
    "            return len(self.return_gRNA_seq())\n",
    "        \n",
    "        return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "772a5b49",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enter filepath of CSV file with genes to search: C:\\Users\\ievut\\OneDrive\\Desktop\\Oncogene Library Project\\Some 30 Genes of Oncogene Library.csv\n",
      "Enter desired filepath to save valid genes and gRNA sequences CSV: C:\\Users\\ievut\\OneDrive\\Desktop\\Oncogene Library Project\\valid_genes_\n",
      "Enter desired filepath to save invalid gene CSV: C:\\Users\\ievut\\OneDrive\\Desktop\\Oncogene Library Project\\invalid_genes_\n",
      "Enter PAM to search for: TTTC\n",
      "Enter repeat sequence: UAAUUUCUACUAAGUGUAGAU\n",
      "Enter length of target sequence: 20\n",
      "Enter species of collection of genes: human\n",
      "querying 1-31...done.\n",
      "Finished.\n",
      "22 input query terms found dup hits:\n",
      "\t[('ACSM2A', 2), ('ACSM2B', 2), ('ACSM3', 2), ('ACSS2', 2), ('ACTA2', 2), ('ABCB1', 2), ('ACTBP2', 2)\n",
      "1 input query terms found no hit:\n",
      "\t['ACT']\n",
      "Finished.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "55it [01:27,  1.58s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "querying 1-31...done.\n",
      "Finished.\n",
      "22 input query terms found dup hits:\n",
      "\t[('ACSM2A', 2), ('ACSM2B', 2), ('ACSM3', 2), ('ACSS2', 2), ('ACTA2', 2), ('ABCB1', 2), ('ACTBP2', 2)\n",
      "1 input query terms found no hit:\n",
      "\t['ACT']\n",
      "Finished.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "55it [01:16,  1.39s/it]\n"
     ]
    }
   ],
   "source": [
    "def main():\n",
    "    current_date_time = datetime.datetime.now()\n",
    "    formatted_curr_date_time = current_date_time.strftime(\"%Y-%m-%d_%Hh-%Mm-%Ss\")\n",
    "\n",
    "    while True:\n",
    "        input_CSV = input(\"Enter filepath of CSV file with genes to search: \")\n",
    "\n",
    "        if os.path.exists(input_CSV):\n",
    "            break\n",
    "        else:\n",
    "            print(\"Invalid filepath. Please try again.\")\n",
    "            \n",
    "    output_valid_CSV_file_path = input(\"Enter desired filepath to save valid genes and gRNA sequences CSV: \") + formatted_curr_date_time + \".csv\"\n",
    "   \n",
    "    output_invalid_CSV_file_path = input(\"Enter desired filepath to save invalid gene CSV: \") + formatted_curr_date_time + \".csv\"\n",
    "\n",
    "    while True:\n",
    "        input_PAM_seq = input(\"Enter PAM to search for: \")\n",
    "\n",
    "        allowed_letters = set(\"ATGC\")\n",
    "        uppercase_input = input_PAM_seq.upper()\n",
    "\n",
    "        if set(uppercase_input) <= allowed_letters:\n",
    "            break\n",
    "        else:\n",
    "            print(\"The input contains characters other than A, T, G, and C. Please try again.\")\n",
    "    \n",
    "    while True:\n",
    "        input_repeat_seq = input(\"Enter repeat sequence: \")\n",
    "\n",
    "        allowed_letters = set(\"AUGC\")\n",
    "        uppercase_input = input_repeat_seq.upper()\n",
    "\n",
    "        if set(uppercase_input) <= allowed_letters:\n",
    "            break\n",
    "        else:\n",
    "            print(\"The input contains characters other than A, U, G, and C. Please try again.\")\n",
    "    \n",
    "    while True:\n",
    "        input_length_of_target_seq = input(\"Enter length of target sequence: \")\n",
    "\n",
    "        try:\n",
    "            input_length_of_target_seq = int(input_length_of_target_seq)\n",
    "            break\n",
    "        except ValueError:\n",
    "            print(\"Please enter an integer.\")\n",
    "      \n",
    "    while True:\n",
    "        input_species = input(\"Enter species of collection of genes: \")\n",
    "        target_gene_library = TargetSeqLibrary(input_CSV, input_species, input_PAM_seq, input_repeat_seq, input_length_of_target_seq, output_valid_CSV_file_path, output_invalid_CSV_file_path)\n",
    "\n",
    "        try:\n",
    "            target_gene_library.check_every_gene_for_target_seq()\n",
    "            break\n",
    "            \n",
    "        except requests.exceptions.HTTPError as http_err:\n",
    "            print(\"Invalid species. Please check all valid species in NCBI taxonomy database.\")  \n",
    "            \n",
    "    target_gene_library = TargetSeqLibrary(input_CSV, input_species, input_PAM_seq, input_repeat_seq, input_length_of_target_seq, output_valid_CSV_file_path, output_invalid_CSV_file_path)\n",
    "    target_gene_library.check_every_gene_for_target_seq()\n",
    " \n",
    "if __name__ == \"__main__\":\n",
    "    main()\n",
    "\n",
    "    "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
